intellij/src/saros/intellij/eventhandler/filesystem/LocalFilesystemModificationHandler.java

```diff
@@ -367,35 +368,103 @@ private void generateResourceDeletionActivity(@NotNull VirtualFileEvent virtualF
       LOG.trace("Reacting before resource deletion: " + deletedVirtualFile);
     }
 
-    SPath path = VirtualFileConverter.convertToSPath(project, deletedVirtualFile);
+    if (deletedVirtualFile.isDirectory()) {
+      generateFolderDeletion(deletedVirtualFile);
+    } else {
+      generateFileDeletion(deletedVirtualFile);
+    }
+  }
+
+  /**
+   * Generates and dispatches a folder deletion activity for the deleted folder.
+   *
+   * @param deletedFolder the folder that was deleted
+   */
+  private void generateFolderDeletion(@NotNull VirtualFile deletedFolder) {
+    SPath path = VirtualFileConverter.convertToSPath(project, deletedFolder);
 
     if (path == null || !session.isShared(path.getResource())) {
       if (LOG.isTraceEnabled()) {
-        LOG.trace("Ignoring non-shared resource deletion: " + deletedVirtualFile);
+        LOG.trace("Ignoring non-shared folder deletion: " + deletedFolder);
       }
 
       return;
     }
 
     User user = session.getLocalUser();
 
-    IActivity activity;
+    Deque<IActivity> queuedDeletionActivities = new ConcurrentLinkedDeque<>();
```

Comment:
Why must this be a concurrent collection ?

Reply:
Because I wasn't sure whether there is any kind of parallelization when iterating the resources with the `ContentIterator`. The interface/method descriptions don't make any statement on this, so I wanted to make sure that there aren't any issues with concurrent writes.